### PR TITLE
[receiver/hostmetrics] Add available memory state (optionally)

### PR DESCRIPTION
**Pull request by lootek**
**Description:** <Describe what has changed.>

Add `available` memory state to hostmetricsreceiver.
Add `enable_additional_memory_states` configuration option to memoryscraper that controls reporting of memory state metrics: `slab_reclaimable`, `slab_unreclaimable` and `available`. This is because when those are enabled, all the memory states don't sum up well (the result exceeds `total` memory).

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7417

**Testing:** <Describe what testing was performed and which tests were added.>
Updated the relevant memoryscraper unit tests.

**Documentation:** <Describe the documentation added.>
The hostmetricsreceiver/memoryscraper docs were updated accordingly.

### CODE REVIEW

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ attributes:
   state:
     description: Breakdown of memory usage by type.
     type: string
-    enum: [buffered, cached, inactive, free, slab_reclaimable, slab_unreclaimable, used]
+    enum: [buffered, cached, inactive, free, slab_reclaimable, slab_unreclaimable, used, available]
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml:9]: Can you please expand the list in multiple lines and add comments for the states that are disabled by default

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/config.go [hunk 2]
```diff
@@ -23,4 +23,10 @@ import (
 type Config struct {
 	metadata.MetricsBuilderConfig `mapstructure:",squash"`
 	internal.ScraperConfig
+
+	// EnableAdditionalMemoryStates (false by default) controls reporting of memory state metrics:
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/config.go:27]:
```suggestion
	// EnableAdditionalMemoryStates (false by default) controls reporting of data points with additional memory states:
```

File: .chloggen/hostmetrics-add-mem-available.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
Comment (dmitryax) [inline .chloggen/hostmetrics-add-mem-available.yaml:2]: Please add another `breaking` change log item for disabling slab states

**Review by dmitryax — COMMENTED** [review]
I think it's a safe solution, but not 100% sure it's sustainable in the long run. 

@rogercoll @djaglowski WDYT?

Another option would be to introduce additional optional metrics like `system.linux.memory.slab` and `system.memory.available`. BTW is the available state value useful on other platforms? If not, it can be `system.linux.memory.available`

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper_linux.go [hunk 4]
```diff
@@ -29,15 +29,23 @@ func (s *scraper) recordMemoryUsageMetric(now pcommon.Timestamp, memInfo *mem.Vi
...
 func (s *scraper) recordMemoryUtilizationMetric(now pcommon.Timestamp, memInfo *mem.VirtualMemoryStat) {
 	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Used)/float64(memInfo.Total), metadata.AttributeStateUsed)
 	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Free)/float64(memInfo.Total), metadata.AttributeStateFree)
 	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Buffers)/float64(memInfo.Total), metadata.AttributeStateBuffered)
 	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Cached)/float64(memInfo.Total), metadata.AttributeStateCached)
-	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Sreclaimable)/float64(memInfo.Total), metadata.AttributeStateSlabReclaimable)
+}
+
+func (s *scraper) recordMemoryUtilizationMetricAdditionalStates(now pcommon.Timestamp, memInfo *mem.VirtualMemoryStat) {
+	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Used)/float64(memInfo.Total), metadata.AttributeStateUsed)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper_linux.go:48]: seems like a bug
```suggestion
	s.mb.RecordSystemMemoryUtilizationDataPoint(now, float64(memInfo.Sreclaimable)/float64(memInfo.Total), metadata.AttributeStateSlabReclaimable)
```

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml (hunk 1 above)
Comment (lootek) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml:9]: good idea  👍

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper_linux.go (hunk 4 above)
Comment (lootek) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/memory_scraper_linux.go:48]: oops, thanks! :)

**Comment by lootek** [comment]
> BTW is the available state value useful on other platforms?

Looks like it is: https://github.com/shirou/gopsutil/blob/c7c548d6bdb46b39cd36ae0815926de70ab1b8ca/mem/mem.go#L22

File: receiver/hostmetricsreceiver/README.md [hunk 5]
```diff
@@ -88,6 +88,15 @@ load:
   cpu_average: <false|true>
 ```
 
+### Memory
+
+`enable_additional_memory_states` specifies whether to enable reporting of a couple of additional memory state metrics: `slab_reclaimable`, `slab_unreclaimable` and `available`. When those are enabled, all the memory states don't sum up well (the result exceeds `total` memory) (default: `false`).
```
Comment (djaglowski) [inline receiver/hostmetricsreceiver/README.md:93]: This isn't a hill I'm going to die on, but in my opinion we should be looking to replace this metric with one or more metrics that use attributes appropriately to differentiate portions of a meaningful total.

I'm not an expert at memory management but it seems someone should be able to clearly state the relationships between these values. As a (probably wrong) starting point:
- `total = buffered + cached + inactive + free + used`
- `slab = slab_reclaimable + slab_unreclaimable`
- `available = free + slab_reclaimable`?

If we had a list of these relationships, it should be fairly easy to identify a correct data model. Without it, I don't know how to evaluate such a solution except to say that overloading a metric has never been necessary elsewhere, as far as I am aware. Again, this isn't intended to be blocking - just trying to articulate _why_ in my opinion we are having to invent novel ways to tweak a metric.

File: receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml (hunk 1 above)
Comment (lootek) [reply receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml]: done

File: .chloggen/hostmetrics-add-mem-available.yaml (hunk 3 above)
Comment (lootek) [reply .chloggen/hostmetrics-add-mem-available.yaml:2]: done

File: receiver/hostmetricsreceiver/README.md (hunk 5 above)
Comment (lootek) [reply receiver/hostmetricsreceiver/README.md:93]: Yeah, you're right, we're dealing with a complex issue that doesn't have a simple and universal solution. However, I believe that including the ability to report available memory is a valuable feature for users, despite its potential drawbacks.

**Comment by rogercoll** [comment]
> Another option would be to introduce additional optional metrics like `system.linux.memory.slab` and `system.memory.available`. BTW is the available state value useful on other platforms? If not, it can be `system.linux.memory.available`

Memory available only has a unique value for Linux systems, as it was introduced by the kernel in version 3.16. For other platforms, it seems that Gopsutils is computing the value from others:

- Windows: Same as Free -> https://github.com/shirou/gopsutil/blob/master/mem/mem_windows.go#L49 (Duplicated metric)
- BSD: Inactive +Cached + Free (we currently provide all the values) -> https://github.com/shirou/gopsutil/blob/master/mem/mem_openbsd.go#L49
- Darwin: Inactive + Free (already provided) -> https://github.com/shirou/gopsutil/blob/master/mem/mem_darwin_cgo.go#L43

I would consider introducing `system.linux` or only provide the metric for Linux systems to prevent duplications. For other platforms, we could add some documentation on how to extract the equivalent value.

**Comment by lootek** [comment]
@rogercoll thank you for sharing your thoughts on this topic! I do agree with some of what you've said, but there's another issue that I think we should consider as well. While it's true that the Linux kernel provides the `available` memory state exclusively, there are also memory states that are specific to other OSes. To me, unifying our approach would require some significant changes to the current memory scraper metrics/codebase. What do you think? Do you agree it'd be better to address this as a separate issue or pull request?

**Comment by lootek** [comment]
@dmitryax @djaglowski what do you think?

**Comment by dmitryax** [comment]
I'd lean toward introducing new optional metrics `system.memory.slab` and `system.linux.memory.available` instead of having optional attributes violating the "aggregation meaningfulness" requirement.

It should provide users with a better source in case they want to utilize both available and total linux memory metrics.

**Comment by lootek** [comment]
> I'd lean toward introducing new optional metrics `system.memory.slab` and `system.linux.memory.available` instead of having optional attributes violating the "aggregation meaningfulness" requirement.
> 
> It should provide users with a better source in case they want to utilize both available and total linux memory metrics.

Fine, I can rewrite my PR according to this then. Just to be sure:
* `system.memory.slab` is going to be a new metric with no differentiation between reclaimable and unreclaimable (having those available from the previous place)
* `system.linux.memory.available` is gonna be reported for linux only? What about the other OSes that have [memory available supported by gopsutil](https://github.com/shirou/gopsutil/blob/v3.21.11/mem/mem.go#L22)?

**Comment by lootek** [comment]
> > `system.memory.slab` is going to be a new metric with no differentiation between reclaimable and unreclaimable (having those available from the previous place). Also, to my understanding slab refers to the memory used by kernel specifically so it should be probably also available for linux (or unix-based) OSes only?
> 
> Right, this should be available on Linux only, but I don't think we need to have `linux` prefix in the name because we don't have a conflicting meaning of `slab` as per [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/system-metrics.md#systemos---os-specific-system-metrics). The metric should have the same state attribute with `reclaimable` and `unreclaimable` values.

So is this gonna duplicate what we currently have? Or replace the existing memory states at some point?

> > `system.linux.memory.available` is gonna be reported for linux only? What about the other unix-based OSes that have [memory available supported by gopsutil](https://github.com/shirou/gopsutil/blob/v3.21.11/mem/mem.go#L22)?
> 
> As pointed out by @rogercoll, it looks like the metric value is unique for Linux only, and other Unix-based OSes like FreeBSD can get that value deferred from other states.

ok, this makes sense. Unless we want to provide memory available for those as a kind of "handy shortcut".

> I'd like to get an agreement on this from @rogercoll and @djaglowski before starting implementation. We should also start by adding `system.memory.slab` to [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/system-metrics.md#systemmemory---memory-metrics) to get more people involved.

Good point, for sure we should!

> Also, these issues should be handled as separate follow-up PRs. Let's keep this one for the discussion.
> 

ok 👍 

> And thank you, @lootek, for helping with that! I think we are getting closer to finally resolving #7417

My pleasure :) And if possible I'd like to take part in having this finally resolved.

**Comment by dmitryax** [comment]
> So is this gonna duplicate what we currently have? Or replace the existing memory states at some point?

We will need to remove `slab_reclaimable` and `slab_unreclaimable` states from `system.memory.usage` metric and move them to `system.memory.slab`.

**Comment by lootek** [comment]
> > So is this gonna duplicate what we currently have? Or replace the existing memory states at some point?
> 
> We will need to remove `slab_reclaimable` and `slab_unreclaimable` states from `system.memory.usage` metric and move them to `system.memory.slab`.

Since such removal would be a breaking change I guess it'll be a typical 2-stage process (first the deprecation, then actiual removal)? Or are we gonna add a config setting like I did here in this PR?

**Comment by rogercoll** [comment]
I would not tie the metrics with the values provided by Gopsutils (the underlying library could be changed), instead, to the ones provided by each kernel. If a user **requires** does metrics to be provided for all the systems by the collector, the user could use the [metrics generation processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricsgenerationprocessor) to generate the new ones (at the moment does not support metric attributes, I can open an issue/pr for that) as they are currently being provided. 

[Specs issue](https://github.com/open-telemetry/opentelemetry-specification/issues/3341) to add the new system.memory.slab metric into the specification, which will solve the first issue of system.memory states not summing up to the total value. Please feel free to add extra information or change the proposal.

**Comment by dmitryax** [comment]
> Since such removal would be a breaking change I guess it'll be a typical 2-stage process (first the deprecation, then actiual removal)? Or are we gonna add a config setting like I did here in this PR?

Right, it should be 2-step approach
1. Add `system.memory.slab` (optional) + add a warning for `system.memory.usage` saying that slab attributes are deprecated and will be removed, use `system.memory.slab` instead.
2. Remove slab attributes from `system.memory.usage`

**Pull request closed without merging**